---
title: "`mga`: User-friendly `R` package for microbial genetic analysis of amplicon
  data"
author: "Erika Y. Lin"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

\center Department of Biology, University of British Columbia Okanagan

BIOL 440: Honours Thesis

Advisor: Dr. Michael J. Noonan

April 24, 2024

<!-- `r Sys.Date()` -->

\newpage
\flushleft

## **Abstract**

Studying microbes is fundamental in advancing a wide array of scientific fields, from ecology to medicine and pharmaceutical research. With the increasing use of amplicon sequencing technology, new bioinformatics tools are needed to process and analyze microbial DNA sequencing data. Existing software for microbiome analysis, although extensive and established, are often either inflexible or difficult to learn for new users. Furthermore, to ensure reproducibility, researchers frequently build their workflows by coding in command-line programs like `R`, an open-source software made for quantitative analysis and graphical visualization. However, available `R` packages target different steps of the DNA analysis process and may be unintuitive for researchers with limited coding experience. To address this lack of rigorous but also easy-to-use software for microbiome analysis, we developed the `R` package `mga`: Microbial Genetic Analysis, that streamlines several key steps of genetic data processing. This openly accessible software package distinguishes genetic sequences, classifies taxonomy, builds evolutionary and co-occurrence networks, and calculates key measures of ecological community structure from filtered DNA sequences. To effectively handle genetic data for any microbial community analysis, the core function `mga()` was designed to be comprehensive, versatile, and user-friendly, collapsing thousands of lines of code into a few simple, yet flexible inputs. `mga` was tested for functionality and sensitivity with fungal and bacterial amplicon sequencing datasets, producing results consistent with the literature, and it is built to handle various microbial taxa targeted for amplicon sequencing. By facilitating downstream analyses, taxonomic filtering, and data visualization, `mga` supports reproducible and accessible microbiome research. 

\newpage
## **Introduction**

Advances in microbiological research over the last 20 years has contributed to growing interest in microbiota among new generations of researchers entering the field. The advent of polymerase chain reaction (PCR) catalyzed the rapid innovation of techniques for research involving microbial genetics and a tandem increase in the application of these techniques towards all fields of biological research. The development of gene sequencing technologies and associated bioinformatics tools for data handling and analysis has been fundamental to overcoming limitations in investigating microorganisms, supporting transformative research in ecology (Coolen et al., 2022; Rotoni et al., 2022; Shade et al., 2012), evolution (Pekkonen et al., 2013; Ward et al., 2021), environmental sciences (Castañeda Alejo et al., 2022; Lahlali et al., 2022), epidemiology (McLean et al., 2022), health sciences (Latorre-Pérez et al., 2021; Turnbaugh et al., 2007), and medical research (Caldara et al., 2022; Percival et al., 2015; Yang et al., 2022). As the technologies and software used continue to progress, it can be difficult to decide on which tools to use for microbiome analyses, particularly for new and early career researchers. Researchers should ideally be working with bioinformatics tools that are accessible, reproducible, convenient to use, and suitable for their study area and question (Wen et al., 2023). Additionally, with the growing population of researchers in microbial genetics, analytical tools should be easy to learn for new users.

|       Current established software can be categorized either as a graphical user interface (GUI) or character user interface (CUI) program, each with their own merits. GUIs, programs with point-and-click graphics, such as the widely used QIIME2 (Bolyen et al., 2019) and mothur (Schloss et al., 2009) applications, are designed to be accessible and easy to use. However, as users do not have access to the backend processes, GUIs have limited flexibility in both the services they provide, as well as customization of specific functions. Moreover, graphics-based tools contribute to the reproducibility crisis in research, due to lack of clarity in documenting analyses, thus making it difficult to replicate any studies done using GUIs.

|       In contrast, command-line CUIs, such as `R` (R Core Team, 2021) and Python, enable reproducible analyses via code availability and are also very flexible with the ability to tailor arguments within functions and to select specific functions among a collection of software packages. The increasing awareness of the need for transparency in scientific research has resulted in a shift favouring command-line applications, such that popular bioinformatics software like QIIME2 and mothur both have CUI plug-in interface options. However, the flexible nature of CUI programs makes them inherently inaccessible and challenging to learn, particularly for researchers with little coding experience. Furthermore, available DNA analysis workflows in `R` are long and thus, can be computationally heavy and difficult to follow, even for experienced researchers. Wen et al. (2023) reviewed 324 `R` packages for microbiome analysis, recognizing that these packages target different parts of the analysis and are suited to specific fields of microbial research. The last decade has seen the emergence of more integrated `R` packages intended for comprehensive microbiome data processing, such as `phyloseq` (McMurdie & Holmes, 2013); however, these packages tend to lack versatility and/or functionality (Wen et al., 2023). Data visualization may also be poorly integrated within packages (Wen et al., 2023) or prone to error.

|       To facilitate accessibility while maintaining reproducibility, we have developed `mga`: Microbial Genetic Analysis, an `R` package that greatly simplifies the complicated process of working with microbial sequences, effectively streamlining key components of the microbiome analysis workflow. With the aim of incorporating the benefits of GUI applications into a CUI-based program, our package is openly accessible, comprehensive, versatile, and user-friendly. Principal steps in sequence data processing, prior to conducting diversity or difference analyses, involve filtering and trimming sequence files, inferring operational taxonomic units (OTUs) or amplicon sequence variants (ASVs) within samples, assigning taxonomy to OTUs or ASVs, and reconstructing the phylogeny of the sample communities (Fig 1; Callahan, Sankaran, et al., 2016). These steps are performed universally for microbiome analyses, regardless of the tools used; as such, the core function `mga()` distills this primary component of the workflow into a short function with simple inputs. The core function merges trimmed sequence files from each sample to identify unique ASVs for taxonomic classification and phylogenetic tree construction, as well as generating co-occurrence networks and associated diagnostics. The output is a `mga-class` object that stores all the data, including a table containing computed measures of community diversity, facilitating downstream analyses. `mga()` was constructed based on the `phyloseq` and `DADA2` workflow by (Callahan, Sankaran, et al., 2016), which leverages the fine resolution of ASV inference over OTU clustering based on fixed but arbitrary dissimilarity thresholds (Callahan et al., 2017).

![Standard microbial DNA analysis workflow from raw genetic sequences to final interpretation of results. (1) Genetic sequence data are filtered to (2) identify unique ASVs or operational taxonomic units (OTUs), (3) compared to a database of known organisms for taxonomic classification. (4) Phylogeny is inferred based on relatedness and genetic similarity. Follow-up steps can differ by field of study or research question.](vignettes/figures/microbiome_workflow.png)
	
|       `mga()` works directly with raw and filtered genetic sequences in the text-based fastq format in which they are obtained from sequencing services, requiring only that the directory storing the extracted files be named in the following manner: "FOL_DER”, such that the file path for any single fastq file is: “FOL_DER/SAMPLENAME_XXX.fastq…”. The main elements needed for the function are the raw forward and reverse sequence reads (`fnFs`, `fnRs`), the filtered forward and reverse sequence reads (`filtFs`, `filtRs`), and the reference library FASTA file used for taxonomic classification (`refFasta`). Sample metadata can also be included using the `metadata` argument, attaching any information provided to the returned outputs. The `mga()` function has additional arguments that can be specified to customize parts of the analysis, such as parameters for adjusting phylogenetic tree and co-occurrence network construction. These arguments ensure that the function is flexible, while building in standard default settings that maintain reasonable outcomes for users unfamiliar or unconcerned with the details of specific arguments.

```{r mga, eval = FALSE}

mga(fastq.Fs = fnFs,  # file paths for forward raw fastq files
    fastq.Rs = fnRs,  # file paths for reverse raw fastq files
    filtFs = filtFs,  # file paths for forward filtered fastq files
    filtRs = filtRs,  # file paths for reverse filtered fastq files
    refFasta = silva.132,  # 16S silva reference FASTA with species assignment
    metadata = meta_data)  # sample metadata (optional)

# All arguments and defaults
mga(fastq.Fs, fastq.Rs,  # file paths for forward and reverse raw fastq files
    filtFs, filtRs,  # file paths for filtered and trimmed sequences
    refFasta,  # file path of reference FASTA for taxonomic classification
    metadata = NULL,  # sample metadata in list or data frame format
    make.unique = TRUE,  # defines all unclassified taxa as being unique
    group.species = TRUE,  # agglomerates phyloseq taxa by species for computed metrics
    tree.args = list(k = 4,  # phylogenetic tree parameters
                     inv = 0.2,
                     model = "GTR",
                     rearrangement = "stochastic"),
    network = TRUE,  # runs network analysis
    network.args = list(type = "taxa",  # co-occurrence network parameters
                        distance = "jaccard", 
                        max.dist = 0.45, 
                        keep.isolates = TRUE),
    seed = 100,  # seed for stochastic processes
    multithread = FALSE,  # set to FALSE on Windows operating systems
    verbose = TRUE)  # relays messages for each step

```

|       We used fungal and bacterial datasets to test the `mga` package for the performance of the main `mga()`, along with additional functions incorporated for subsequent analyses (`drop_taxa()`) and data visualization (`plot.mga`). A sensitivity analysis was also conducted to test for the package’s sensitivity to a subjective user input, such as choice of reference library for taxonomic classification. These analyses have been adapted into vignettes, available with supplementary analyses on GitHub at: https://github.com/ErikaYLin/mga_Microbial-Genetic-Analysis. The package can be installed from GitHub at: https://github.com/ErikaYLin/mga. 

## **Methods and Materials**

### *The `mga()` function:*

In developing `mga`, we used both bacterial and fungal DNA sequencing files. We used a published dataset of highly-overlapping Illumina Miseq 2x250 16S region amplicon bacterial sequencing files accessed from the workflow written by Callahan, Sankaran, et al. (2016), originally collected by Schloss et al. (2012) for their study on the stabilization of the murine gut microbiome post weaning. ITS region amplicon sequences for fungal data were taken from an unpublished study on the effects of cover crops on soil fungi in Okanagan vineyards (unpublished data). Usage of `mga()` is demonstrated as follows using a subset of the data collected by Schloss et al. (2012).

|       The core function `mga()` is adapted from the “Bioconductor Workflow for Microbiome Analysis” by Callahan, Sankaran, et al. (2016), in which sequences are processed from raw reads to `phyloseq-class` objects then used for data visualization and graphical analyses. `mga()` starts from the sequence dereplication step to combine replicate sequences and remove redundant information, reducing computation time (Callahan, Sankaran, et al., 2016). The DADA2 method is employed to differentiate between sequencing errors from genetic variation and to independently infer sequence variants, eliminating substitution and indel errors (Callahan, Sankaran, et al., 2016): 

```{r dada2, eval = FALSE}

# Dereplicating identical sequences
derepFs <- dada2::derepFastq(filtFs, verbose = TRUE)
derepRs <- dada2::derepFastq(filtRs, verbose = TRUE)
# Naming the derep-class objects by the sample names
names(derepFs) <- sampleNames
names(derepRs) <- sampleNames

# Estimating sequencing error rates pre-DADA2
# The application of the DADA2 method follows the workflow by Callahan et al. (2017).
# https://bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/Microbiome/MicrobiomeWorkflowII.html
errF <- dada2::learnErrors(filtFs, multithread = multithread)
errR <- dada2::learnErrors(filtRs, multithread = multithread)

# DADA2 method to infer ASVs
dadaFs <- dada2::dada(derepFs, err = errF, multithread = multithread)
dadaRs <- dada2::dada(derepRs, err = errR, multithread = multithread)

```

|       Sequence pairs output from `dada2` (v1.28.0; Callahan, McMurdie, et al., 2016) processing are paired and chimeras are removed to build a table of ASVs, which are fed into the `assignTaxonomy()` function from the `phyloseq` (v1.44.0; McMurdie & Holmes, 2013) package for taxonomic classification, using the user provided reference library. Any unidentified taxa are renamed as unclassified taxa following a method by Hui (2021). By default, these unclassified taxa are all considered unique (numbered), but can be kept as is by switching the `make.unique` argument to FALSE.

```{r ASVtab, eval = FALSE}

# Merging sequence pairs
merged <- dada2::mergePairs(dadaFs, derepFs, dadaRs, derepRs)
# Constructing sequence table
seqtabAll <- dada2::makeSequenceTable(merged)
# Remove chimeras from the sequence table
ASVtab <- dada2::removeBimeraDenovo(seqtabAll)

# Assign taxonomy to the ASVs with chosen reference FASTA and build a taxonomy table
taxTab <- dada2::assignTaxonomy(ASVtab,
                                refFasta = refFasta,
                                multithread = multithread)

```

|       To construct phylogenetic trees for each sample, the `DECIPHER` (v2.28.0; Wright, 2016) package is used for multiple sequence alignment. `phangorn` (v2.11.1; Schliep, 2011) is the main package used in building phylogenies; for each sample, and initial tree is created using the neighbour-joining method, however specific parameters and methods for clustering and tree arrangement can be customized using `tree.args`. Here we use the default settings: `k = 4, inv = 0.2, model = “GTR”, rearrangement = “stochastic”`. The full list of possible models and rearrangement types used for tree construction can be found at the manual file for the `phangorn::optim.pml()` function.

```{r phylotree, eval = FALSE}

# Multiple sequence alignment
seqs <- dada2::getSequences(ASVtab)
names(seqs) <- seqs # sequences as tree tip labels
alignment <- DECIPHER::AlignSeqs(Biostrings::DNAStringSet(seqs), anchor=NA, verbose=FALSE)

# Phylogenetic tree construction
phangAlign <- phangorn::phyDat(methods::as(alignment, "matrix"), type = "DNA")
dm <- phangorn::dist.ml(phangAlign)
treeNJ <- phangorn::NJ(dm) # initial neighbour-joining tree
fit = phangorn::pml(treeNJ, data = phangAlign)
fitGTR <- stats::update(fit, k = tree.args$k, inv = tree.args$inv)
fitGTR <- phangorn::optim.pml(fitGTR, model = tree.args$model,
                              optInv = TRUE, optGamma = TRUE,
                              rearrangement = tree.args$rearrangement,
                              control = phangorn::pml.control(trace = 0))

```

|       The ASV table, taxonomic table, phylogenetic tree, and sample metadata are stored in a `phyloseq-class` object for easier handling in subsequent steps. All elements in the `phyloseq-class` object are agglomerated by species for downstream analyses, unless otherwise specified (`group.species = FALSE`). 

```{r phyloseq, eval = FALSE}

# Combine data into single phyloseq-class object
tax = phyloseq::tax_table(as.matrix(taxTab))
ps <- phyloseq::phyloseq(phyloseq::otu_table(ASVtab,
                                             taxa_are_rows = FALSE),
                         phyloseq::sample_data(sample_info),
                         tax,
                         phyloseq::phy_tree(fitGTR$tree))

# Aggregate species
species <- phyloseq::tax_glom(ps, taxrank = 'Species', NArm = FALSE)

```

|       The `phyloseq` package is employed for measures of community structure, including species richness, alpha diversity (Shannon’s H and Simpson’s diversity index), and phylogenetic diversity (phylogenetic distance). 

```{r diversity, eval = FALSE}

# Sum presences in each sample for species richness
rich <- nrow(phyloseq::tax_table(species))
# total individual sequences from all species in each sample
n <- apply(ASVtab, 1, function(x) sum(x))
ASVs <- ncol(ASVtab) # total number of ASVs

# Alpha species diversity measures
# Shannon index
ps_alpha_div <- phyloseq::estimate_richness(if (group.species){species} else{ps},
                                            split = TRUE, measure = "Shannon")
# Simpson index
ps_alpha_div2 <- phyloseq::estimate_richness(if (group.species){species} else{ps},
                                             split = TRUE, measure = "Simpson")

# Phylogenetic diversity (sum tree branch lengths)
PD <- sum(phyloseq::tree_layout(
  phyloseq::phy_tree(if (group.species){species} else{ps}))$edgeDT[["edge.length"]])

```

|       Additionally, the `make_network()` function is used to generate a co-occurrence network for each sample based on a threshold ecological distance between vertices (species or ASVs) that determines whether vertices are more likely to co-occur together (connected) or not (unconnected). The option to bypass co-occurrence network and network parameters can be adjusted in the `mga()` arguments. Default network parameters used here are: `type = “taxa”, distance = “jaccard”, max.dist = 0.45, keep.isolates = TRUE`. Network diagnostics are calculated using the `igraph` (v1.4.2; Csárdi et al., 2006) package to count the total number of vertices ($v$) and edges ($e$), the degree of each vertex, the network connectivity ($e/v$), and network connectance ($e/v^2$; Landi et al., 2018). 

### *Package Performance & Sensitivity Analysis:*

To test for the package’s sensitivity to the user’s choice of reference library, we ran the `mga()` function for the bacterial dataset using three different 16S reference libraries selected from two commonly used databases for bacterial metabarcoding: Silva and GreenGenes. An older Silva library (16S Silva NR v132 training set; Callahan, 2018) and an updated version (16S Silva NR99 v138.1 training set with species; McLaren & Callahan, 2021) were used to assess the importance of using new libraries. From the GreenGenes database, the 16S GreenGenes v13.8 training set was used (Callahan, 2016). See the Appendix for the full code for the data processing and analyses.

|       Sample metadata and amplicon sequences for this study were obtained from the link provided in the workflow by Callahan, Sankaran, et al. (2016) and processed prior to being input into the core function. Sequencing quality was visually verified using the `plotQualityProfile()` function in the `dada2` package for filtering and trimming. Upon inspecting the first two sequence reads, forward reads sustained relatively high average quality and were trimmed at position 240; reverse read quality decreased substantially around position 160 and thus, these were trimmed at position 160 (Appendix; Callahan, Sankaran, et al., 2016). The first ten base pairs were also trimmed to avoid potential pathological issues as discussed by Callahan, Sankaran, et al. (2016). The raw fastq files were filtered and trimmed with the `filterAndTrim()` function from `dada2` for a maximum of 2 expected errors per read.

|       The raw and filtered sequence files, as well as each reference library and the metadata were inputted into the `mga()` function (example below). The default settings were used for all other arguments not shown. The analysis was run once for each reference library, looping over each sample file and results were stored in a list. Results for each sample were extracted and merged into a common data frame with added meta data. The phylogenetic diversity for each sample was compared across all three reference libraries to determine differences in outputted results.

```{r bacteria, eval = FALSE}

# Empty list for storing the results
SILVA.OLD <- list()

# Loop the analysis process for every file in the directory
for (i in 1:length(fnFs)){

  # Assign metadata to be input as a list
  meta_data <- list()
  for (j in 1:length(fnFs)){

    ID <- sapply(strsplit(fnFs[i], "_"), `[`, 2)
    ID <- sapply(strsplit(ID, "/"), `[`, 2)
    meta_data[[j]] <- as.list(meta[meta$sample.ID %in% ID,])
  }

  # Create a new row of results for each output from `mga`
  SILVA.OLD[[i]] <- mga::mga(fastq.Fs = fnFs[i],
                            fastq.Rs = fnRs[i], 
                            filtFs = filtFs[i],
                            filtRs = filtRs[i],
                            refFasta = silva.132,
                            metadata = meta_data[[j]])
}

```

|       Package functionality and performance were also tested for fungal sequencing data, using ITS fungal data from an unpublished study on soil fungi (unpublished data) and the UNITE general FASTA release library for ITS fungal data (Abarenkov et al., 2022). Furthermore, as the length of an analysis may be of concern to users, the computation time per sample for `mga()` was measured with the `tictoc` package (v1.2.1; Izrailev, 2024).

### *Taxonomic Filtering & Data Visualization:*

The `drop_taxa()` function in the `mga` package prunes `mga-class` objects for select taxa of interest. This function operates similarly to the `mga()` function with an added matching and filtering process, facilitated by the `prune_taxa()` function from the `phyloseq` package. The primary inputs for the function are the `mga-class` object produced by `mga()` and a two-column table of the species to keep. The table was formatted such that taxa names were added to a column titled “taxon” and the taxonomic level to which each listed taxon belongs filling a second “group” column. We used a subset of unpublished ITS fungal data collected for a study on the soil fungi of different cover crops to demonstrate the use of the `drop_taxa()` function in identifying potential pathogens. A table of known fungal plant pathogens (Bekris et al., 2021; Travadon et al., 2022; Úrbez-Torres et al., 2014) and default network parameters were used for taxonomic filtering. See Appendix for the full-code example.

```{r drop_taxa, eval = FALSE}

# Empty list to store filtered mga objects
test_filter <- list()
# Filter each object in the combined output list from `mga()`
for (i in 1:length(UNITE.fungi)) {
  test_filter[[i]] <- mga::drop_taxa(UNITE.fungi[[i]],
                                     taxa = pathogens)  # table of select taxa
}

```

|       Furthermore, the `mga` package includes a wrapper function `plot.mga()` for `phyloseq` plotting techniques. This function allows simple data visualization of `mga` outputs, including the phylogenetic tree and co-occurrence network. We showed example uses of this plotting function for the analysis results of the bacterial dataset (Schloss et al., 2012) to generate figures of the phylogenetic tree for the second sample by specifying `type = “tree”` and the co-occurrence network for the second sample by specifying `type = “network”`. Graphical parameters for the tree and network were adjusted based on arguments from the `plot_tree()` and `plot_network()` functions in the `phyloseq` package.

```{r plotting, eval = FALSE}

# Plot phylogenetic tree for sample 2 with the Silva NR99 v138.1 library
plot(SILVA.NEWS[[2]], type = "tree", cex = 0.25) # smaller tip label font

# Plot co-occurrence network
plot(SILVA.NEWS[[2]], type = "network",
     point_size = 2, label = NULL) # remove labels, shrink points

```

## **Results**

```{r packages_setup, echo = FALSE, warning = FALSE, message = FALSE}

# Load requisite packages
library(ggplot2)
library(rphylopic)
library(cowplot)
library(mga)

# Load mga results and saved data frame
load(file = "vignettes/RDS/silva132_mga_results.rda")
results.df <- readRDS("vignettes/RDS/results.df_silva132_mga.rds")
load(file = "vignettes/RDS/silva138.1S_mga_results.rda")
results.df2 <- readRDS("vignettes/RDS/results.df_silva138.1S_mga.rds")
load(file = "vignettes/RDS/gg13.8_mga_results.rda")
results.df3 <- readRDS("vignettes/RDS/results.df_gg13.8_mga.rds")
load(file = "vignettes/RDS/unite.its_mga_results.rda")
results.df4 <- readRDS("vignettes/RDS/results.df_unite.its_mga.rds")

```

### *Package Performance & Outputs:*

Overall, the package performed well when tested with both bacterial and fungal amplicon data using different targeted regions, including the 16S rRNA gene for prokaryotes and the ITS region for fungal taxa. For each sample that was fed into the core function, `mga()` outputted a `mga-class` list storing all data produced during sequence processing and analysis. The resulting list contained a frequency table of all identified sequences, including chimeras (seqtabAll); a frequency table of  unique ASVs (ASVtab); the corresponding taxonomy table (taxTab); the models and methods used to construct the phylogenetic tree; the sample metadata if provided (sampledata), the `phyloseq` object (ps) used to store the ASV table (`otu_table` format), taxonomy table (`tax_table` format), phylogenetic tree (`phy_tree` format), and sample data (`sample_data` format); the `phyloseq` object aggregated by species (ps.species); a data frame of diversity and network measures merged with metadata (results.samples); a co-occurrence network if specified (net); and a data frame tabulating the degree of each vertex. Computed diversity metrics were species alpha diversity (Shannon’s H and Simpson’s diversity index), species richness as the total species count, phylogenetic diversity (summed phylogeny branch lengths), total read counts for all ASVs combined (n.indivs), and total ASV count. Network diagnostics were calculated for the total number of vertices and edges, network connectivity as the average degree per vertex or interactions per species, and network connectance as the proportion of edges present of the total possible number of edges (Landi et al., 2018). Below presents an example of the `mga-class` output for a single sample. See Appendix for full examples and code.

|       The computation time was proportional to the number of samples processed, increasing linearly as samples were added to the analysis (Fig A4). While a linear increase in run time for the main function is preferred over an exponential increase, which would greatly hinder processing ability, we recognize that there are certain steps within the `mga()` function that may be responsible for slowing the analysis process.

```{r results1, echo = FALSE, message = FALSE, warning = FALSE}

# Objects in stored in each `mga` object
n1 <- t(data.frame(names(SILVA.NEWS[[1]])[c(1:5)], names(SILVA.NEWS[[1]])[c(6:10)]))
rownames(n1) <- NULL
knitr::kable(n1, col.names = NULL, 
             caption = "Objects stored in a `mga` object for an individual sample.")

# Sample results
n2 <- head(results.df2[,c(1:11)])
rownames(n2) <- NULL
knitr::kable(n2,  # Results for silva NR99 v138.1
             caption = "Diversity measures and network diagnostics for the 
             first six samples in the mga-class object produced using the 
             Silva NR99 v138.1 reference library.")
```

### *Sensitivity to Reference Libraries:*

The results produced by `mga()` were similar across reference libraries, indicating that the package was insensitive to this subjective user input. Specifically, although individual measures of diversity and network topology showed differences between reference libraries, the overall conclusions drawn from diversity metrics, such as phylogenetic diversity, were consistent regardless of the library used (Fig 2). The updated Silva v138.1 library produced phylogenetic diversity values for each sample that were more similar to the GreenGenes v13.8 library than to the older Silva v132 library.

```{r results2, echo = FALSE, message = FALSE, warning = FALSE, fig.height = 3.5, fig.width = 5.5, fig.cap = "Boxplot of the phylogenetic diversity of murine gut bacteria in the days immediately following weaning (0-9 days) and months later (141-150 days). Results are compared after evaluating the analysis using three different reference libraries for taxonomic classification."}

# Remove row names from results data frame
div.metrics <- results.df[,c(1:5,27)]
rownames(div.metrics) <- NULL
div.metrics2 <- results.df2[,c(1:5,27)]
rownames(div.metrics2) <- NULL
div.metrics3 <- results.df3[,c(1:5,27)]
rownames(div.metrics3) <- NULL

# Extract diversity indices and dpw group for each ref library
PD.df <- list(Silva_old = div.metrics,
              Silva_new = div.metrics2,
              Greengenes = div.metrics3)

# Add ref libraries
PD.df$Silva_old$ref <- rep("Silva_132", 19)
PD.df$Silva_new$ref <- rep("Silva_138.1", 19)
PD.df$Greengenes$ref <- rep("Greengenes_13.8", 19)

# Collapse into data frame
PD.df <- do.call(rbind, PD.df)
rownames(PD.df) <- NULL # remove row names

# Boxplot of PD across different ref libraries
PD_box <- ggplot(data = PD.df, aes(x = dpwgroup, y = PD)) +
  geom_boxplot(size = 0.5, aes(fill = dpwgroup)) +
  # scale_fill_manual(values = palette3) +
  geom_jitter(shape = 1, width = 0.2) +
  facet_wrap(~ref) +
  labs(fill = NULL, x = "Days Post Weaning", y = "Phylogenetic Diversity") +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.title.y = element_text(size = 12, family = "sans", face = "bold", hjust = 0.5),
        axis.text.y = element_text(size = 10, family = "sans"),
        axis.text.x = element_text(size = 10, family = "sans"),
        axis.title.x = element_text(size = 12, family = "sans", face = "bold", vjust = 1),
        strip.text = element_text(size = 12, family = "sans"),
        legend.position = "none",
        panel.background = element_blank(),
        panel.border = element_rect(linewidth = 0.2, fill = NA),
        plot.background = element_blank(),
        plot.margin = unit(c(0.2,0.1,0,0.2), "cm")) # top, right, bottom, left
PD_box

```

![(A) Bar plot of the relative abundance of species within each fungal class for buckwheat and buffalo grass samples including all species initially identified. (B) Bar plot of the relative abundance of plant pathogens of interest after pruning for pathogenic fungal taxa. (C) Bar plot comparing the relative abundance of pathogenic to non-pathogenic species.](vignettes/figures/abundance_barplot_all.png)

### *Taxonomic Filtering & Data Visualization:*

The `drop_taxa()` function was able to successfully prune the `mga` object for the pathogenic taxa of interest appointed by the table of taxa and taxonomic levels provided to the function. The resulting output contained only the pathogenic taxa, allowing the comparison of the relative abundances of pathogenic to non-pathogenic taxa (Fig 3C), as well as the relative abundance of different groups of pathogens (Fig 3B). On average, the most abundant taxon of fungal pathogens found in soil for both buckwheat and buffalo grass cover crops was the Nectriaceae family (Fig 3C). This could be attributed to the Nectriaceae being a higher taxonomic level under which there are many species. 

|       The plotting function was also able to facilitate visualization of the phylogenetic tree and co-occurrence networks for each sample analyzed. The aesthetic parameters of the tree and network plots were able to be modified according to graphical parameters from the `phyloseq` `plot_tree()` and `plot_network()` functions. As examples, using the Silva v138.1 bacterial analysis results, the tree tip label font sizes were adjusted (Fig A5) or labels were completely removed (Fig 4), and the network node sizes were adjusted (Fig 5).

```{r datavis, warning = FALSE, message = FALSE, fig.cap = "Phylogenetic tree for sample F3D1 using the Silva v138.1 training FASTA. Branch tip labels were removed for cleaner visualization.", echo = FALSE, fig.height = 5, fig.width = 5}

plot(SILVA.NEWS[[2]], type = "tree", show.tip.label = FALSE) # remove tip labels

```

```{r datavis2, warning = FALSE, message = FALSE, fig.cap = "Co-occurrence network for sample F3D1 using the Silva v138.1 training FASTA. The vertex sizes were adjusted and labels were removed.", echo = FALSE, fig.height = 3, fig.width = 3.5}

# Plot co-occurrence network
plot(SILVA.NEWS[[2]], type = "network", point_size = 2, label = NULL)
# remove labels and shrink points

```

|       The network plot appeared different each time plotted, however, although the spacing between individual vertex points changed, the edge lengths joining vertices remains the same. Thus, the network connectivity and overall topology were unchanged for the same sample.

\newpage
## **Discussion**

To facilitate microbial genetic analyses and sequence processing, we created the `mga` package, addressing the need for accessible and easily operable tools. By developing the package for the `R` programming language, `mga` was made available to the expanding body of researchers in biology who use command-line programs for statistical research, in recognition the importance of maintaining reproducibility and flexibility in their studies. Taking a step back from specialized but increasingly complex tools, `mga` contributes a user-friendly alternative into the pool of contemporary bioinformatics tools. This package addresses the three pillars upon which the Open Science framework was built: transparency in research, accessibility and sharing, and inclusivity for researchers of at all stages of their careers (Center for Open Science, 2024).

|       Using published bacterial sequencing files (Schloss et al., 2012) and fungal sequencing files, we constructed the package for amplicon data and evaluated its functionality. In testing package performance, we aggregated results by species, based on the default setting (`group.species = TRUE`). It is important to note that if results had not been agglomerated by species (`group.species = FALSE`), individual ASVs would have been used to calculate diversity measures and to construct the co-occurrence network. For example, the total number of network vertices can be equivalent to either the species count or ASV count, depending on function specifications; each vertex either represents a unique species or ASV. However, some ASVs may correspond to the same species, thus, unless this is a point of interest for a user’s specific research, we recommend keeping the default argument.

|       The sensitivity analysis conducted on the choice of reference library and database showed comparable diversity values across libraries. The findings from this analysis, that murine gut microbiota stabilizes over time after weaning from initial communities with greater variability and ecological diversity, were consistent with literature (Schloss et al., 2012), supporting package performance, and findings did not vary significantly between reference libraries, indicating that the package was not sensitive to this subjective input. This was desirable because although the reference FASTA training sets used for assigning taxonomy differ both among databases and between updates due to the continued growth in knowledge of microbial systems, the dominant findings and conclusions should remain robust, particularly if there have not been major changes in the understanding of microbial evolution. While the taxonomic classification step of `mga()` was found to be insensitive to the user input, this does not discount the possibility for other subjective parameters in phylogenetic modelling and network construction to impact findings for specific cases and types of data. The uncomplicated design of the `mga()` function and its flexible arguments enables users to test the sensitivity of parameters for their own data, by rerunning the function with incremental changes to numerical inputs. The package uses established standard settings for arguments, however, users unfamiliar with parameters involved in key steps of the process have the capacity to perform sensitivity analyses for these arguments to ensure that reasonable results are obtained prior to follow-up work.

|       As presented through the bacterial and fungal case studies, `mga` products can be readily employed in downstream analyses. These often include diversity and differential analyses to characterize and describe microbial communities. Here, we have demonstrated how diversity metrics calculated as part of the `mga()` analysis, such as phylogenetic diversity, can be either be directly plotted to visualize preliminary results or fed into ensuing models for further investigation. As metabarcoding is subject to PCR amplification and sequencing bias (Krehenwinkel et al., 2017), absolute read counts are often unreliable, and thus researchers commonly compare relative abundances within and across samples. Relative abundance bar plots were generated to demonstrate taxonomic classification and filtering results; in using the `mga` outputs from the fungal analysis, the workflow to create these widely used plotting techniques was made straightforward. As all data are stored within the `mga-class` object returned by the `mga()` and `drop_taxa()` functions, this allows for convenient handling in establishing a centralized unit containing all necessary components for a variety of subsequent analyses.

|       Naturally, we recognize that the current package does have limitations, as do all software. First and foremost, manual filtering and trimming of sequence files is required prior to inputting files as `mga()` arguments. Due to sequencing quality and error rates varying by case and by company, it would be both difficult and inappropriate to standardize this step of microbiome analysis. As such, users are encouraged to use the initial steps from the workflow by Callahan, Sankaran, et al. (2016) to sort through fastq files in the designated file path, using the `plotQualityProfile()` function from the `dada2` package to visually determine positions where average sequencing quality decreases, and to filter and trim sequence files based on the quality profiles. As `dada2` and `phyloseq` packages are integral components of `mga`, our package is able to smoothly handle any outputs and feed into any functions from either `dada2` or `phyloseq`. However, with the integration of `phyloseq-class` objects into `mga`, certain outputs within the `mga-class` list produced from the core function can only be extracted using `phyloseq`-specific syntax and functions. Although elements stored in `phyloseq-class` format (ASV table, taxonomy table, phylogenetic tree) are S4 class objects and thus, cannot be retrieved via typical data frame and list extraction methods (i.e. `$`, `[]`, or `[[]]`), this format effectively organizes the elements for access to a wider range of downstream analyses, including those that make use of `phyloseq` syntax. Therefore, this limitation trades off with increased flexibility in follow-up work.

|       We seek to increase the efficiency of the core function in future updates to the package, including incorporating global parallelization to reduce computation times. Moreover, in accordance with our goal of convenience and accessibility, the package has undergone rigorous testing and modification to ensure that it is suitable for CRAN publication. While the package is publicly available on GitHub for installation, publication within CRAN, a community-vetted repository, would allow easier installation from directly within `R`, recognizing that `mga` is well-designed by strict standards. Altogether, despite minor limitations, the `mga` package leverages existing `R` packages and workflows to form a comprehensive, yet easy-to-use bioinformatics tool for microbial genetic sequence processing. Not only does this package serve as a versatile, user-oriented software, helping to introduce new researchers in microbiology to code-based analytical tools, the development of `mga` also acts as a call for greater accessibility in science and a framework for building user-friendly analytical software in all fields of research.

\newpage
## **Acknowledgments**

I would like to thank Dr. Michael Noonan for his guidance throughout the research process and for overseeing the project. Thank you also to Dr. Miranda Hart, Dr. Daniel Rosa, Dr. Mehdi Sharifi, and Selina Spence for providing me with the fungal dataset with which I was able to develop and test the package. Additionally, a published bacterial dataset by Schloss et al. (2012) was also used to support the development of this package. Finally, I am grateful for all the support I have received from the members of the Quantitative Ecology Lab both within and beyond my research.

\newpage
## **References**
\setlength{\parindent}{-0.4in}
\setlength{\leftskip}{0.4in}
\noindent

Abarenkov, K., Zirk, A., Piirmann, T., Pöhönen, R., Ivanov, F., Nilsson, R. H., & Kõljalg, U. (2022). UNITE general FASTA release for Fungi. *UNITE Community*. https://dx.doi.org/10.15156/BIO/2483911

Bekris, F., Vasileiadis, S., Papadopoulou, E., Samaras, A., Testempasis, S., Gkizi, D., Tavlaki, G., Tzima, A., Paplomatas, E., Markakis, E., Karaoglanidis, G., Papadopoulou, K. K., & Karpouzas, D. G. (2021). Grapevine wood microbiome analysis identifies key fungal pathogens and potential interactions with the bacterial community implicated in grapevine trunk disease appearance. *Environmental Microbiome, 16*(1), 23. https://doi.org/10.1186/s40793-021-00390-1

Bolyen, E., Rideout, J. R., Dillon, M. R., Bokulich, N. A., Abnet, C. C., Al-Ghalith, G. A., Alexander, H., Alm, E. J., Arumugam, M., Asnicar, F., Bai, Y., Bisanz, J. E., Bittinger, K., Brejnrod, A., Brislawn, C. J., Brown, C. T., Callahan, B. J., Caraballo-Rodríguez, A. M., Chase, J., … Caporaso, J. G. (2019). Reproducible, interactive, scalable and extensible microbiome data science using QIIME 2. *Nature Biotechnology, 37*(8), 852–857. https://doi.org/10.1038/s41587-019-0209-9

Caldara, M., Belgiovine, C., Secchi, E., & Rusconi, R. (2022). Environmental, Microbiological, and Immunological Features of Bacterial Biofilms Associated with Implanted Medical Devices. *Clinical Microbiology Reviews, 35*(2). https://doi.org/10.1128/cmr.00221-20

Callahan, B. J. (2016). The RDP and GreenGenes taxonomic training sets formatted for DADA2 [Data set]. *Zenodo*. https://doi.org/10.5281/zenodo.158955

Callahan, B. J. (2018). Silva taxonomic training data formatted for DADA2 (Silva version 132) [Data set]. *Zenodo*. https://doi.org/10.5281/zenodo.1172783

Callahan, B. J., McMurdie, P. J., & Holmes, S. P. (2017). Exact sequence variants should replace operational taxonomic units in marker-gene data analysis. *The ISME Journal, 11*(12), 2639–2643. https://doi.org/10.1038/ismej.2017.119

Callahan, B. J., McMurdie, P. J., Rosen, M. J., Han, A. W., Johnson, A. J. A., & Holmes, S. P. (2016). DADA2: High-resolution sample inference from Illumina amplicon data. *Nature Methods, 13*(7), 581–583. https://doi.org/10.1038/nmeth.3869

Callahan, B. J., Sankaran, K., Fukuyama, J. A., McMurdie, P. J., & Holmes, S. P. (2016). Bioconductor Workflow for Microbiome Data Analysis: from raw reads to community analyses. *F1000Research, 5*, 1492. https://doi.org/10.12688/f1000research.8986.2

Castañeda Alejo, S. M., Tejada Meza, K., Valderrama Valencia, M. R., Arenazas Rodríguez, A. J., & Málaga Espinoza, C. J. (2022). Tire Ground Rubber Biodegradation by a Consortium Isolated from an Aged Tire. *Microorganisms, 10*(7), 1414. https://doi.org/10.3390/microorganisms10071414

Center for Open Science. (2024). *Open Science*. Center for Open Science. https://www.cos.io/open-science

Coolen, S., Rogowska-van der Molen, M., & Welte, C. U. (2022). The secret life of insect-associated microbes and how they shape insect–plant interactions. *FEMS Microbiology Ecology, 98*(9). https://doi.org/10.1093/femsec/fiac083

Csárdi, G., Nepusz, T., Traag, V., Horvát, S., Zanini, F., Noom, D., & Müller, K. (2006). The igraph software package for complex network research. *InterJournal*, *Complex Systems*, 1695. https://doi.org/10.5281/zenodo.7682609

Hui, Y. (2021, January 1). *Tutorial for microbiome analysis in R*. Yan Hui. https://www.yanh.org/2021/01/01/microbiome-r/

Izrailev, S. (2024, March 18). *tictoc: Functions for Timing R Scripts, as Well as Implementations of “Stack” and “StackList” Structures*. CRAN. https://cran.r-project.org/package=tictoc

Krehenwinkel, H., Wolf, M., Lim, J. Y., Rominger, A. J., Simison, W. B., & Gillespie, R. G. (2017). Estimating and mitigating amplification bias in qualitative and quantitative arthropod metabarcoding. *Scientific Reports, 7*(1), 17668. https://doi.org/10.1038/s41598-017-17333-x

Lahlali, R., Ezrari, S., Radouane, N., Kenfaoui, J., Esmaeel, Q., El Hamss, H., Belabess, Z., & Barka, E. A. (2022). Biological Control of Plant Pathogens: A Global Perspective. *Microorganisms, 10*(3), 596. https://doi.org/10.3390/microorganisms10030596

Landi, P., Minoarivelo, H. O., Brännström, Å., Hui, C., & Dieckmann, U. (2018). Complexity and stability of ecological networks: a review of the theory. *Population Ecology, 60*(4), 319–345. https://doi.org/10.1007/s10144-018-0628-3

Latorre-Pérez, A., Hernández, M., Iglesias, J. R., Morán, J., Pascual, J., Porcar, M., Vilanova, C., & Collado, L. (2021). The Spanish gut microbiome reveals links between microorganisms and Mediterranean diet. *Scientific Reports, 11*(1), 21602. https://doi.org/10.1038/s41598-021-01002-1

McLaren, M. R., & Callahan, B. J. (2021). Silva 138.1 prokaryotic SSU taxonomic training data formatted for DADA2 [Data set]. *Zenodo*. https://doi.org/10.5281/zenodo.4587955

McLean, G., Kamil, J., Lee, B., Moore, P., Schulz, T. F., Muik, A., Sahin, U., Türeci, Ö., & Pather, S. (2022). The Impact of Evolving SARS-CoV-2 Mutations and Variants on COVID-19 Vaccines. *MBio, 13*(2). https://doi.org/10.1128/mbio.02979-21

McMurdie, P. J., & Holmes, S. (2013). phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data. *PLoS ONE, 8*(4), e61217. https://doi.org/10.1371/journal.pone.0061217

Pekkonen, M., Ketola, T., & Laakso, J. T. (2013). Resource Availability and Competition Shape the Evolution of Survival and Growth Ability in a Bacterial Community. *PLoS ONE, 8*(9), e76471. https://doi.org/10.1371/journal.pone.0076471

Percival, S. L., Suleman, L., Vuotto, C., & Donelli, G. (2015). Healthcare-associated infections, medical devices and biofilms: risk, tolerance and control. *Journal of Medical Microbiology, 64*(4), 323–334. https://doi.org/10.1099/jmm.0.000032

R Core Team. (2023). *R: A Language and Environment for Statistical Computing* (4.3.0). R Foundation for Statistical Computing. https://www.R-project.org

Rotoni, C., Leite, M. F. A., Pijl, A., & Kuramae, E. E. (2022). Rhizosphere microbiome response to host genetic variability: a trade-off between bacterial and fungal community assembly. *FEMS Microbiology Ecology, 98*(6). https://doi.org/10.1093/femsec/fiac061

Schliep, K. P. (2011). phangorn: Phylogenetic analysis in R. *Bioinformatics, 27*(4), 592–593. https://doi.org/10.1093/bioinformatics/btq706

Schloss, P. D., Schubert, A. M., Zackular, J. P., Iverson, K. D., Young, V. B., & Petrosino, J. F. (2012). Stabilization of the murine gut microbiome following weaning. *Gut Microbes, 3*(4), 383–393. https://doi.org/10.4161/gmic.21008

Schloss, P. D., Westcott, S. L., Ryabin, T., Hall, J. R., Hartmann, M., Hollister, E. B., Lesniewski, R. A., Oakley, B. B., Parks, D. H., Robinson, C. J., Sahl, J. W., Stres, B., Thallinger, G. G., Van Horn, D. J., & Weber, C. F. (2009). Introducing mothur: Open-Source, Platform-Independent, Community-Supported Software for Describing and Comparing Microbial Communities. *Applied and Environmental Microbiology, 75*(23), 7537–7541. https://doi.org/10.1128/AEM.01541-09

Shade, A., Peter, H., Allison, S. D., Baho, D. L., Berga, M., Bürgmann, H., Huber, D. H., Langenheder, S., Lennon, J. T., Martiny, J. B. H., Matulich, K. L., Schmidt, T. M., & Handelsman, J. (2012). Fundamentals of Microbial Community Resistance and Resilience. *Frontiers in Microbiology, 3*. https://doi.org/10.3389/fmicb.2012.00417

Travadon, R., Lawrence, D. P., Moyer, M. M., Fujiyoshi, P. T., & Baumgartner, K. (2022). Fungal species associated with grapevine trunk diseases in Washington wine grapes and California table grapes, with novelties in the genera Cadophora, Cytospora, and Sporocadus. *Frontiers in Fungal Biology, 3*. https://doi.org/10.3389/ffunb.2022.1018140

Turnbaugh, P. J., Ley, R. E., Hamady, M., Fraser-Liggett, C. M., Knight, R., & Gordon, J. I. (2007). The Human Microbiome Project. *Nature, 449*(7164), 804–810. https://doi.org/10.1038/nature06244

Úrbez-Torres, J. R., Haag, P., Bowen, P., & O’Gorman, D. T. (2014). Grapevine trunk diseases in British Columbia: Incidence and characterization of the fungal pathogens associated with black foot disease of grapevine. *Plant Disease, 98*(4), 456–468. https://doi.org/10.1094/PDIS-05-13-0524-RE

Ward, R. D., Stajich, J. E., Johansen, J. R., Huntemann, M., Clum, A., Foster, B., Foster, B., Roux, S., Palaniappan, K., Varghese, N., Mukherjee, S., Reddy, T. B. K., Daum, C., Copeland, A., Chen, I.-M. A., Ivanova, N. N., Kyrpides, N. C., Shapiro, N., Eloe-Fadrosh, E. A., & Pietrasiak, N. (2021). Metagenome Sequencing to Explore Phylogenomics of Terrestrial Cyanobacteria. *Microbiology Resource Announcements, 10*(22). https://doi.org/10.1128/MRA.00258-21

Wen, T., Niu, G., Chen, T., Shen, Q., Yuan, J., & Liu, Y.-X. (2023). The best practice for microbiome analysis using R. *Protein & Cell, 14*(10), 713–725. https://doi.org/10.1093/procel/pwad024

Wright, E. S. (2016). Using DECIPHER v2.0 to Analyze Big Biological Sequence Data in R. *The R Journal, 8*(1), 352–359. https://doi.org/10.18129/B9.bioc.DECIPHER

Yang, J., Long, H., Hu, Y., Feng, Y., McNally, A., & Zong, Z. (2022). Klebsiella oxytoca Complex: Update on Taxonomy, Antimicrobial Resistance, and Virulence. *Clinical Microbiology Reviews, 35*(1). https://doi.org/10.1128/CMR.00006-21

